### PR TITLE
Introduce `ResettableHandler` interface and `attach()` method

### PR DESCRIPTION
**Pull request by TahaZahid05**
Partially Addresses #3535 

**Description:**
Introduces a `ResettableHandler` abstract base class to establish a formalized interface for stateful handlers. `EarlyStopping` is the first adopter of this pattern, implementing both `reset()` and `attach()` methods. 

This mirrors the existing `Metric.attach()` lifecycle pattern, avoiding any Engine-level modifications while providing a clean, self-contained way to manage handler state across runs.

**API**

```python
from ignite.engine import Engine, Events
from ignite.handlers import EarlyStopping

trainer = Engine(train_fn)
evaluator = Engine(eval_fn)

handler = EarlyStopping(patience=5, score_function=score_fn, trainer=trainer)

# Attaches the handler's check to the evaluator (Events.EPOCH_COMPLETED)
# AND automatically attaches the handler's `reset()` to the trainer (Events.STARTED)
handler.attach(
    engine=evaluator, 
    event=Events.EPOCH_COMPLETED, 
    reset_engine=trainer, 
    reset_event=Events.STARTED
)
```

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

**Comment by TahaZahid05** [comment]
@vfdev-5 I have refactored the PR to use the `Metric.attach()` pattern as we discussed.

**Key changes:**
- Reverted all modifications to `Engine` (`engine.py` and `events.py` are now untouched).
- Added the `ResettableHandler` ABC in `ignite/base/mixins.py` with an abstract `reset()` method to establish the formalized interface.
- Refactored `EarlyStopping` to inherit from the ABC, implementing `reset()` and `attach()`. 
- Added test coverage in `test_early_stopping.py` verifying that `attach()` registers both the evaluation event and the reset callback (including cross-engine scenarios).
- Sphinx documentation configuration updated to recognize the new ABC.

Let me know what you think!

**Future Direction:**
If you are happy with how this pattern turned out for `EarlyStopping`, I'd be more than happy to help migrate the other stateful handlers in follow-up PRs. Just let me know if that's something you'd want!

**Comment by vfdev-5** [comment]
@TahaZahid05 can you please also update the PR description. Thanks a lot for working on this topic!

File: ignite/base/mixins.py [hunk 1]
```diff
@@ -1,7 +1,21 @@
...
 
+class ResettableHandler(metaclass=ABCMeta):
+    """Interface for handlers whose internal state can be reset.
+
+    Subclasses must implement the :meth:`reset` method to clear any accumulated
+    state, typically at the beginning of a training run.
+    """
+
+    @abstractmethod
+    def reset(self) -> None:
```
Comment (vfdev-5) [inline ignite/base/mixins.py:20]: I wonder whether this interface should also have `attach` method? 
Should it be abstractclass or a [Protocol](https://typing.python.org/en/latest/spec/protocol.html)?

File: ignite/handlers/early_stopping.py [hunk 2]
```diff
@@ -128,6 +128,38 @@ def __call__(self, engine: Engine) -> None:
...
+        """Reset the early stopping state, including the counter and best score."""
+        self.counter = 0
+        self.best_score = None
+
+    def attach(
+        self,
+        engine: Engine,
+        event: Any = Events.COMPLETED,
+        reset_engine: Engine | None = None,
+        reset_event: Any = Events.STARTED,
```
Comment (vfdev-5) [inline ignite/handlers/early_stopping.py:144]: Here we can also think about HandlerUsage similar to [MetricUsage](https://github.com/pytorch/ignite/blob/18c719344f0a4b9ce7198079803cecebd48d009b/ignite/metrics/metric.py#L30) and maybe even unify both into a single class...

Also, we can promote `attach` method as alternative to direct adding the handler to the engine in the sense that, it still works like `trainer.add_event_handler(event, handler)`, but if we care about handler's state lifecycle during the training we can use `attach` method. So attach does the reset by default according to the default usage.

File: ignite/base/mixins.py (hunk 1 above)
Comment (TahaZahid05) [reply ignite/base/mixins.py:20]: I considered adding an `attach` method to the interface, but the method signature differs significantly across handlers. To enforce a universal signature on the abstract class would force us to use `def attach(self, engine: Engine, *args: Any, **kwargs: Any) -> None:` which may be less effective as it removes the type-hinting/auto-complete for the users. Do you think we should still add it?

Regarding using `Protocol` or `ABCMeta`, I chose `ABCMeta` because all major interfaces (`Metric`, `Checkpoint`, etc) in the repository are using it so I wanted to align the function with the rest of the repository. I didn't see `Protocol` being used anywhere in the codebase.

File: ignite/handlers/early_stopping.py (hunk 2 above)
Comment (TahaZahid05) [reply ignite/handlers/early_stopping.py:144]: Yeah, that would be great. I can work on that. Creating a unifying single class like `Usage` (open to suggestions on the name)  would require modifying base files, changing docs, and imports across the codebase. Would you prefer I scope that unificiation into a dedicated follow-up PR, or should I attempt to refactor `MetricUsage` alongside `EarlyStopping` within this PR?

Comment (TahaZahid05) [reply ignite/handlers/early_stopping.py:144]: Hey @vfdev-5, following up on your feedback about generalizing `MetricUsage` to a shared `Usage`

I've opened a dependent PR here: #3579 

Since the new PR relies on the `attach()` logic we introduced in this one, we can review and merge this PR first. Once merged, I will rebase the new PR so it merges cleanly. Let me know what you think!

File: ignite/base/mixins.py (hunk 1 above)
Comment (vfdev-5) [reply ignite/base/mixins.py:20]: Even if `def attach(self, engine: Engine, *args: Any, **kwargs: Any) -> None:` is less informative, it is still an interface asking to override `attach` method = essential thing we want to do. What do you think?

File: ignite/base/mixins.py [hunk 3]
```diff
@@ -1,7 +1,21 @@
...
 from collections import OrderedDict
 from collections.abc import Mapping
 
 
+class ResettableHandler(metaclass=ABCMeta):
+    """Interface for handlers whose internal state can be reset.
+
+    Subclasses must implement the :meth:`reset` method to clear any accumulated
+    state, typically at the beginning of a training run.
+    """
```
Comment (vfdev-5) [inline ignite/base/mixins.py:17]: let's put versionadded tag in the docs. Please take a look at the contribution.md guidelines, docs writing part. I think this class should be also exposed in the docs for the users. let's make sure we can see it in the docs.

File: ignite/handlers/early_stopping.py (hunk 2 above)
Comment (vfdev-5) [reply ignite/handlers/early_stopping.py:144]: Sounds good to me

File: ignite/base/mixins.py (hunk 3 above)
Comment (TahaZahid05) [reply ignite/base/mixins.py:17]: Apologies, I'll make sure to be mindful of that from next time.

Comment (vfdev-5) [reply ignite/base/mixins.py:17]: No worries, it happens rather frequently to everyone

File: ignite/base/mixins.py (hunk 1 above)
Comment (TahaZahid05) [reply ignite/base/mixins.py:20]: Yeah, makes sense. I'll change it.

File: ignite/base/mixins.py (hunk 3 above)
Comment (TahaZahid05) [reply ignite/base/mixins.py:17]: @vfdev-5 i have added versionadded tag

**Review by vfdev-5 — APPROVED** [review]
LGTM, thanks @TahaZahid05 !

**Comment by TahaZahid05** [comment]
@vfdev-5 can you guide me regarding the errors? I built it locally and ran fine, and here it seems to be some error regarding `Sampler`, `Dataset`, `DataLoader` but i didn't touch those.

**Comment by vfdev-5** [comment]
"Build docs" errors are failing unrelated. Pyrefly are related to the change but we should avoid circular import of Engine

**Comment by TahaZahid05** [comment]
@vfdev-5 

Thanks. Should I just use a `TYPE_CHECKING` block to import `Engine` and then use a string i.e `engine: "Engine"` in the attach method?

**Comment by vfdev-5** [comment]
> @vfdev-5
> 
> Thanks. Should I just use a `TYPE_CHECKING` block to import `Engine` and then use a string i.e `engine: "Engine"` in the attach method?

Try that and see if it helps

**Comment by TahaZahid05** [comment]
> > @vfdev-5
> > Thanks. Should I just use a `TYPE_CHECKING` block to import `Engine` and then use a string i.e `engine: "Engine"` in the attach method?
> 
> Try that and see if it helps

Done. Ran locally as well and it passed. I have also retained type hints for `attach()` by using `# type: ignore[override]` as i saw it had been implemented in other functions such as for `ProgressBar` in  `tqdm_logger.py` as well.

Thanks.

File: ignite/handlers/early_stopping.py [hunk 4]
```diff
@@ -128,6 +128,45 @@ def __call__(self, engine: Engine) -> None:
...
 
+    def reset(self) -> None:
+        """Reset the early stopping state, including the counter and best score.
+
+        .. versionadded:: 0.6.0
+        """
+        self.counter = 0
+        self.best_score = None
+
+    def attach(  # type: ignore[override]
```
Comment (vfdev-5) [inline ignite/handlers/early_stopping.py:139]: pyrefly fails without `ignore[override]` ?

Comment (TahaZahid05) [reply ignite/handlers/early_stopping.py:139]: @vfdev-5 yes, if ran without `ignore[override]` it gives:

```
ERROR Class member `EarlyStopping.attach` overrides parent class `ResettableHandler` in an inconsistent manner [bad-override]
   --> ignite/handlers/early_stopping.py:139:9
    |
139 |     def attach(  
    |         ^^^^^^
    |
  `EarlyStopping.attach` has type `BoundMethod[EarlyStopping, (self: EarlyStopping, engine: Engine, event: Any = Events.COMPLETED, reset_engine: Engine | None = None, reset_event: Any = Events.STARTED, *args: Any, **kwargs: Any) -> None]`, which is not assignable to `BoundMethod[EarlyStopping, (self: EarlyStopping, engine: Engine, *args: Any, **kwargs: Any) -> None]`, the type of `ResettableHandler.attach`
 INFO 1 error (106 suppressed)
```
